### DISCUSSION TITLE
Misunderstanding of after_flush events?

### DISCUSSION THREAD

**Discussion by fschulze** [sqlalchemy/sqlalchemy#11293] [Usage Questions]
I get sporadic ``DetachedInstanceError`` in a ``after_flush`` handler. How can this happen? My understanding is, that the session isn't committed at that point yet.

I have an attribute listener for ``set`` like this:

```python
@event.listens_for(MyClass.my_attribute, 'set')
def receive_set(my_instance, newvalue, oldvalue, initiator):
    if oldvalue == newvalue:
        return
    session = Session()
    session.info.setdefault('_my_info', []).append((my_instance, oldvalue, newvalue))
```

Then I have an ``after_flush`` listener like this:
```python
@event.listens_for(Session, 'after_flush')
def receive_after_flush(session, flush_context):
    my_infos = session.info.pop('_my_info', [])
    if not my_infos:
        return
    ...
    for (my_instance, oldvalue, newvalue) in my_infos:
        session.add(AuditTableObject(...))
```

In the construction of ``AuditTableObject`` part above I access lazy attributes of ``my_instance`` and sometimes I get the ``DetachedInstanceError`` on one of those. The whole thing is meant to store infos in an audit table if the attribute changed.

Is my understanding of ``after_flush`` wrong and the session might already be commited at this point?

Are there circumstances where the ``set`` event can be triggered by SQLAlchemy itself instead of my code? The code path is relatively short, but I can't find a place where the attribute is set.

I'm quite stumped at how to investigate this, as I might misunderstand some things and it only happens sporadically and I can't reproduce it.

**Comment by zzzeek** [comment]
I'd have to look at the code to see that you wouldn't be in a rollback scenario, where pending Objects would have been evicted.  Objects can be evicted from the session when they were marked deleted also. However I don't think these things should happen in that event .     Your issue may be gc related , or hash ordering related.  Id try to get a reproducer perhaps by running a test repeatedly, set PYTHONHASHSEED= random , etc.

Am traveling so can't look at source code right now

**Comment by fschulze** [comment]
It indeed seems to be from a rollback, I see the following in the traceback info:

```
File sqlalchemy/orm/session.py, line 3584, in _flush
        except:
            with util.safe_reraise():
                transaction.rollback(_capture_exception=True)
```

In case of rollback I don't need to update the audit table, what is the easiest way to detect that from the ``after_flush`` event? I guess something on the ``session`` or ``flush_context``? I couldn't find anything obvious from the docs on ``Session``, ``UOWTransaction`` or ``SessionTransaction`` to determine the current state.

```python
@event.listens_for(Session, 'after_flush')
def receive_after_flush(session, flush_context):
```

Maybe there is a better event chain to achieve my goal?

**Reply by zzzeek** [reply]
have looked at the code.  after_flush is not called after there was a rollback.   if an exception occurs and enters the rollback block you reference above before after_flush() was called, then after_flush() is not called.  If an exception occurs after the after_flush() event, you were not in a rollback when after_flush() was called.     would need a fully runnable no-dependences [MCVE](http://stackoverflow.com/help/mcve) illustrating your actual goal
